### PR TITLE
Interaction Event System

### LINKED ISSUE #87: Interaction system
Includes intents

- There could be multiple ways a held item could interact with a thing
- Interactions could have different 'priorities', to determine which is done by default
- Context menus?
- How do intents work into this. As *beep suggested, do we want to just use help and harm intents?

### PR DESCRIPTION
**Pull request by Dyhr**
## Purpose
This system adds components to trigger interactions between anything in the game. It uses an event based architecture with a very minimal implementation. It supports dependencies between events, such that some events may block others.

## Changes to existing Unity files
- `Assets/Prefab/Furniture/Machinery/locker.prefab`: Added new interactions
- `Assets/Prefab/Human/Human.prefab`: Added interactor that actually sends events
- `Assets/Prefab/Items/Functional/Tools/crowbar.prefab`: Added new interactions

## Technical implementation notes
The core of the system is the `IInteraction` interface that defines the methods needed to receive events, and `InteractionReceiver` which actually receives the events and passes them on to the interactions.
An abstract class simply named `Interaction` inherits from `ScriptableObject` and can be added on to `InteractionReceiver` components to easily add simple interactions to an object. This is mostly intended for fully or mostly stateless interactions, as they aren't attached to the object itself.
For stateful interactions (such as opening a door), it is recommended to extend from `MonoBehaviour`, implement the `IInteraction` interface and require a `InteractionReceiver` sibling component.
It is however possible to implement the interface from anything and there is no requirements that events actually originate from a GameObject. An example extension of this system would be to have a developer console, where one can input events directly to debug and simulate certain situations or to drive admin commands. Any event can be sent to anything from anywhere, so a door doesn't need to care if it is opened by a player, a signal, or an admin command.

## Known issues
- Not everything uses this new system yet

Closes #87

### CODE REVIEW

File: Assets/Scripts/Interaction/Openable.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        public void Setup(Action<string> listen, Action<string> blocks)
+        {
+            listen("open");
+        }
+
+        public bool Handle(InteractionEvent e)
+        {
+            open = !open;
+            
+            Debug.Log($"{name} is now {(open ? "open" : "closed")}");
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Interaction/Openable.cs]: Actual code to open/close in the game world will go here in the future?

File: Assets/Scripts/Interaction/PlaceableSurface.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    public class PlaceableSurface : MonoBehaviour, IInteraction
+    {
+        public void Setup(Action<string> listen, Action<string> blocks)
+        {
+            listen("place");
+        }
+
+        public bool Handle(InteractionEvent e)
+        {
+            var item = e.sender.GetComponent<Item>();
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Interaction/PlaceableSurface.cs]: Check for null and log warning

File: Assets/Scripts/Interaction/PlayerInteractor.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            if (!isLocalPlayer) Destroy(this);
+            hands = GetComponent<Hands>();
+        }
+
+        protected virtual void Update()
+        {
+            if (!camera) camera = Camera.main;
+            if (!camera) return;
+            
+            if (Input.GetMouseButtonDown(0))
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Interaction/PlayerInteractor.cs]: Remove hardcoded button, use input mapping.

File: Assets/Scripts/Interaction/Storage.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        private void Start()
+        {
+            container = GetComponent<Container>();
+            openable = GetComponent<Openable>();
+        }
+
+        public void Setup(Action<string> listen, Action<string> blocks)
+        {
+            listen("store");
+            blocks("open");
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Interaction/Storage.cs]: It does not seem like it is possible to block multiple types of events. Won't we need the ability to do that once we define enough interactions?

File: Assets/Scripts/Interaction/Storage.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+        public bool Handle(InteractionEvent e)
+        {
+            if (onlyIfOpen)
+            {
+                if (openable == null) Debug.LogWarning("Trying to check for a Openable that has not been added");
+                else if (!openable.open) return false;
+            }
+            
+            var item = e.sender.GetComponent<Item>();
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Interaction/Storage.cs]: Check for null

File: Assets/Scripts/Inventory/Container.cs [hunk 6]
```diff
@@ -89,6 +89,7 @@ public virtual void AddItem(int slot, GameObject item)
             throw new Exception("Item already exists in slot"); // TODO: Specific exception
 
         items[slot] = item;
+        item.GetComponent<Item>().container = this;
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Inventory/Container.cs]: On phone, can't see whole script. Null safe?

File: Assets/Scripts/Inventory/Custom/Hands.cs [hunk 7]
```diff
@@ -4,16 +4,14 @@
 
 namespace Inventory.Custom
 {
-    /**
-     * The hand system is an interaction system which connects to the inventory.
-     * There is a hands container, that the interaction system connects to, that defines what items the player is currently holding.
-    */
+    /// <summary>
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Inventory/Custom/Hands.cs]: Cleanup

File: Assets/Scripts/Inventory/Custom/Hands.cs [hunk 8]
```diff
@@ -22,41 +20,31 @@ public class Hands : Interaction, Tool
...
         public int selectedHand = 0;
         public event OnHandChange onHandChange;
-
-        /**
-     * The default hand interaction when no object is present.
-     * Note: This could be moved into a different class if this one gets too cluttered
-     */
-        public void Interact(RaycastHit hit, bool secondary)
+        
+        /// <summary>
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Inventory/Custom/Hands.cs]: Also cleanup

File: Assets/Scripts/Inventory/Custom/Hands.cs [hunk 9]
```diff
@@ -22,41 +20,31 @@ public class Hands : Interaction, Tool
...
-                Container container = gameObject.GetComponent<Container>() ?? gameObject.GetComponentInParent<Container>();
-                if(container)
-                    CmdConnectContainer(container.gameObject);
+                inventory.CmdAddItem(target, handContainer.gameObject, handSlots[selectedHand]);
             }
             else
             {
-                // Drop the item currently being held.
-                inventory.CmdPlaceItem(handContainer.gameObject, handSlots[selectedHand], hit.point + new Vector3(0f, 0.2f), new Quaternion());
+                Debug.LogWarning("Trying to pick up with a non-empty hand");
```
Comment (Em3rgencyLT) [inline Assets/Scripts/Inventory/Custom/Hands.cs]: Is item on item interaction handled or going to be handled somewhere else?

File: Assets/Scripts/Interaction/Openable.cs (hunk 1 above)
Comment (Dyhr) [reply Assets/Scripts/Interaction/Openable.cs]: Yeah, actually working lockers are outside the scope of this

File: Assets/Scripts/Interaction/PlayerInteractor.cs (hunk 3 above)
Comment (Dyhr) [reply Assets/Scripts/Interaction/PlayerInteractor.cs]: This is the same code as we had before. We need to agree on how input should be managed first

File: Assets/Scripts/Interaction/Storage.cs (hunk 4 above)
Comment (Dyhr) [reply Assets/Scripts/Interaction/Storage.cs]: Yes, you can block as many events as you want

File: Assets/Scripts/Inventory/Custom/Hands.cs (hunk 9 above)
Comment (Dyhr) [reply Assets/Scripts/Inventory/Custom/Hands.cs]: Item on item interactions are handled by events, by that when you use an item, you send an event to your held item, which may then forward events to the thing being interacted with

File: Assets/Scripts/Interaction/PlayerInteractor.cs (hunk 3 above)
Comment (Dyhr) [reply Assets/Scripts/Interaction/PlayerInteractor.cs]: This is the same code as we had before. We need to agree on how input should be managed first

File: Assets/Scripts/Interaction/Core/InteractionEvent.cs [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    /// <summary>
+    /// This struct contains all the information about an interaction event.
+    /// </summary>
+    [Serializable]
+    public struct InteractionEvent
+    {
+        /// <summary>
+        /// The kind of event. For example, "pickup", "shoot", "explode", "admin_kill".
+        /// </summary>
+        public string kind;
```
[the remark this answers is not in the document]
Comment (Dyhr) [reply Assets/Scripts/Interaction/Core/InteractionEvent.cs]: Yes, my solution is very controversial. I think my core objection to enums is that I don't like hard coding values.
Additionally strings can be more dynamic if a need presents itself

**Pull request merged**